cmake/AddCxxFlag.cmake

```diff
@@ -2,18 +2,90 @@
 # See LICENSE.txt for details.
 
 # Checks if a CXX flag is supported by the compiler and adds it if it is
+# - FLAG_TO_CHECK: the CXX flag add if the compiler supports it
 function(check_and_add_cxx_flag FLAG_TO_CHECK)
-  include(CheckCXXCompilerFlag)
-  unset(CXX_FLAG_WORKS CACHE)
-  # In order to check for a -Wno-* flag in gcc, you have to check the
-  # -W* version instead.  See http://gcc.gnu.org/wiki/FAQ#wnowarning
-  string(REGEX REPLACE ^-Wno- -W FLAG_TO_CHECK_POSITIVE ${FLAG_TO_CHECK})
-  set(CMAKE_REQUIRED_QUIET 1)
-  check_cxx_compiler_flag(${FLAG_TO_CHECK_POSITIVE} CXX_FLAG_WORKS)
-  unset(CMAKE_REQUIRED_QUIET)
-  if (CXX_FLAG_WORKS)
+  string(REGEX REPLACE ^-Wno- -W POSITIVE_FLAG_TO_CHECK ${FLAG_TO_CHECK})
+  execute_process(
+    COMMAND
+    bash -c
+    "${CMAKE_CXX_COMPILER} -Werror ${POSITIVE_FLAG_TO_CHECK} -x c++ -c \
+- <<< \"\" -o /dev/null"
+    RESULT_VARIABLE RESULT
+    ERROR_VARIABLE ERROR_FROM_COMPILATION
+    OUTPUT_QUIET)
+  if(${RESULT} EQUAL 0)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${FLAG_TO_CHECK}" PARENT_SCOPE)
-  endif()
+  endif(${RESULT} EQUAL 0)
+endfunction()
+
+# Checks which of the CXX FLAGS_TO_CHECK are supported by the compiler and
+# adds them. If adding many flags, this will be much faster than
+# calling check_and_add_cxx_flag multiple times.
```

Comment:
For my own education: is the speedup due to reducing the number of calls to the compiler?

Reply:
There are a couple factors:
1. CMake's checking system basically sets up a mini CMake project with CMake files and a source file. This does a lot of disk activity so just reducing the disk activity helps. That's the idea with streaming in an "empty file" (the escaped quotes in the command).
2. Reducing the number of compiler calls speeds things up. I'm not sure where the overhead comes from here, if it's starting a shell and launching the compiler in it, or launching the compiler, or the compiler itself, but it does seem intuitive that process "all at once" would be better than "one at a time" :)

Does that answer your question?

Reply:
Yes, thanks! 👍